pkg/controller/infrastructure/infraflow/ensure.go

```diff
@@ -163,12 +166,76 @@ func (fctx *FlowContext) ensureKubernetesRoutesCleanup(ctx context.Context) erro
 		return err
 	}
 
-	if err = infrastructure.DeleteRoutes(ctx, fctx.computeClient, cloudRoutes); err != nil {
-		// scale CCM back to originalScale in case if there are errors from DeleteRoutes() call
-		_ = scaleCCMto(originalScale)
+	routes := []v1alpha1.Route{}
+	// Safely retrieve and cast the routes object
+	if routesObj := fctx.whiteboard.GetObject(ObjectKeyRoutes); routesObj != nil {
+		if castedRoutes, ok := routesObj.([]v1alpha1.Route); ok {
+			routes = castedRoutes
+		}
+	}
+
+	// Ensure all cloudRoutes are added to the whiteboard, avoiding duplicates
+	for _, route := range cloudRoutes {
+		instanceName, zone, err := extractInstanceAndZone(route.NextHopInstance)
+		if err != nil {
+			return err
+		}
+
+		// Check if the route already exists in the list
+		exists := false
+		for _, r := range routes {
+			if r.InstanceName == instanceName && r.DestinationCIDR == route.DestRange && r.Zone == zone {
+				exists = true
+				break
+			}
+		}
+
+		// Add the route if it doesn't already exist
+		if !exists {
+			routes = append(routes, v1alpha1.Route{
+				InstanceName:    instanceName,
+				DestinationCIDR: route.DestRange,
+				Zone:            zone,
+			})
+		}
+	}
+	fctx.whiteboard.SetObject(ObjectKeyRoutes, routes)
+	err = fctx.persistState(ctx)
+	if err != nil {
+		fctx.log.Error(err, "failed to persist state")
 	}
 
-	return err
+	// Delete all cloudRoutes in parallel
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(cloudRoutes))
+
+	for _, route := range cloudRoutes {
+		wg.Add(1)
+		go func(route *compute.Route) {
+			defer wg.Done()
+			fctx.log.Info("Deleting route", "name", route.Name, "destRange", route.DestRange)
+			if err := fctx.computeClient.DeleteRoute(ctx, route.Name); err != nil {
+				errChan <- fmt.Errorf("failed to delete route %s: %w", route.Name, err)
+			}
+		}(route)
+	}
+
+	// Wait for all deletions to complete
+	wg.Wait()
+	close(errChan)
+
+	// Collect errors, if any
+	var combinedErr error
+	for err := range errChan {
+		combinedErr = fmt.Errorf("%v; %w", combinedErr, err)
+	}
+
+	// Scale CCM back to originalScale if there were errors
+	if combinedErr != nil {
+		_ = scaleCCMto(originalScale)
+		return combinedErr
+	}
+	return nil
```

Comment:
if we fail the first time executing this function, won't subsequent executions consider the original CCM scale to be 0 ?

Do we really need the scale up or we are okay with the ccm coming up when the CP reconciles.

Reply:
Once we have deleted a route the node won't receive any more traffic until the a new route was created by ccm. So I think it is good to scale ccm up. Or how long will it take until a new CP reconcile is happening? Shall we scale it to one if originalScale is 0?

Reply:
Won't it start receiving traffic as soon as the alias IP is in place ?

Reply:
Yes but we can only attach an aliasIPs to the NIC when all routes were deleted.

Reply:
Then maybe I am misunderstanding something here - please entertain me 😅. 

We scale CCM down, delete all routes and add the secondary range to the subnet. So far, so good.

Now let's assume we scale the CCM up. Questions:
- Can we now create routes on top of the alias IP without issues ?
- If yes, then what happens after the CP reconciliation. In the dual-stack case if I remember correctly we disable the route creation. So we will left over uncleaned routes from enabling the CCM here that would not be managed by the CCM anymore.
   - Is my hypothesis so far correct ?
   - Would the existing routes cause issues as nodes roll out ?

Tldr; we migrate the infrastructure (subnet in this case) but the CCM would be scaled up with the old settings. What can we expect to go wrong there ?

Reply:
We only scale ccm up when not all routes where deleted in which case we have no alias IPs. Once we have alias IPs we have 0 routes and won't reach this part of code anymore as we are directly return nil if there are 0 routes.

Reply:
Ah okay that part is clear now thank you.

Though we lose all progress of the migration in case of an error. Say for example that because of a huge cluster with hundreds of nodes, we hit the context timeout. The subsequent retries will be an arms race with the CCM.

Reply:
As we delete the routes in parallel we are really fast, in my test cases the deletion of all routes took not more than 10 to 12 seconds. I've tried to provoke rate limiting or throttling on gcp side with multiple bulk deletions with 200 routes. I never experienced any of this. If you have a better idea how to do this let me know.